packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,329 @@
+ *                                     elements
+ *
+ * @return {Frame} The stack frame tracking parse progress.
+ */
+function Frame(
+	element,
+	tokenStart,
+	tokenLength,
+	prevOffset,
+	leadingTextStart,
+) {
+	return {
+		element,
+		tokenStart,
+		tokenLength,
+		prevOffset,
+		leadingTextStart,
+		children: [],
+	};
+}
+
+/**
+ * This function creates an interpolated element from a passed in string with
+ * specific tags matching how the string should be converted to an element via
+ * the conversion map value.
+ *
+ * @example
+ * For example, for the given string:
+ *
+ * "This is a <span>string</span> with <a>a link</a> and a self-closing
+ * <CustomComponentB/> tag"
+ *
+ * You would have something like this as the conversionMap value:
+ *
+ * ```js
+ * {
+ *     span: <span />,
+ *     a: <a href={ 'https://github.com' } />,
+ *     CustomComponentB: <CustomComponent />,
+ * }
+ * ```
+ *
+ * @param {string}  interpolatedString  The interpolation string to be parsed.
+ * @param {Object}  conversionMap       The map used to convert the string to
+ *                                      a react element.
+ * @throws {TypeError}
+ * @return {WPElement}  A wp element.
+ */
+const createInterpolateElement = ( interpolatedString, conversionMap ) => {
+	indoc = interpolatedString;
+	offset = 0;
+	output = [];
+	stack = [];
+	tokenizer.lastIndex = 0;
+
+	if ( ! isValidConversionMap( conversionMap ) ) {
+		throw new TypeError(
+			'The conversionMap provided is not valid. It must be an object with values that are WPElements'
+		);
+	}
+
+	do {
+		// twiddle our thumbs
+	} while ( proceed( conversionMap ) );
+	return createElement( Fragment, null, ...output );
+};
+
+/**
+ * Validate conversion map.
+ *
+ * A map is considered valid if it's an object and every value in the object
+ * is a WPElement
+ *
+ * @private
+ *
+ * @param {Object} conversionMap  The map being validated.
+ *
+ * @return {boolean}  True means the map is valid.
+ */
+const isValidConversionMap = ( conversionMap ) => {
+	const isObject = typeof conversionMap === 'object';
+	const values = isObject && Object.values( conversionMap );
+	return isObject &&
+		values.length &&
+		values.every( ( element ) => isValidElement( element ) );
+};
+
+/**
+ * This is the iterator over the matches in the string.
+ *
+ * @private
+ *
+ * @param {Object} conversionMap The conversion map for the string.
+ *
+ * @return {boolean} true for continuing to iterate, false for finished.
+ */
+function proceed( conversionMap ) {
+	const next = nextToken();
+	const [ tokenType, name, startOffset, tokenLength ] = next;
+	const stackDepth = stack.length;
+	const leadingTextStart = startOffset > offset ? offset : null;
+	if ( ! conversionMap[ name ] ) {
+		if ( stackDepth !== 0 ) {
+			const { leadingTextStart: stackLeadingText, tokenStart } = stack.pop();
+			output.push( indoc.substr( stackLeadingText, tokenStart ) );
+		}
+		addText();
+		return false;
+	}
```

Comment:
was this block added to pass a failing test?

I have some questions about it but don't fully understand where it came from or why

Reply:
This block wasn't changed:

```js
addText();
return false;
```

What was _added_ was the block right above it:

```js
if ( stackDepth !== 0 ) {
	const { leadingTextStart: stackLeadingText, tokenStart } = stack.pop();
	output.push( indoc.substr( stackLeadingText, tokenStart ) );
}
```

Yes, it was added because of a failing test.  The failing test was `'returns string without unbalanced tags'` which is testing the string `'This is a <em>string'` and expecting `This is a string`.  But, I just realized this doesn't handle if there are multiple unbalanced tags.  For example `This is <a>a <em>string` ends up as `a &lt;em&gt;string`.  Not good - basically what I added only works if the last tag processed in the string is an unbalanced tag - every other combination will result in an equally mangled string as it does without this code block.   Further, I don't think we can detect these unbalanced tags with the existing algorithm (we'd have to lookahead to see if there is a closing tag and if there isn't either remove the tag or just leave it unparsed).

The question is, should we care? There's two scenarios where this will surface:

### 1. The developer botches writing the original string.

In this case, the botched string will become quite obvious in testing the developers work (and likely picked up in code reviews.

### 2. A botched translated string is subbed in at runtime.

This will still get picked up, but more likely by the end-user of that translation.  The reality however is that this is already a scenario that could happen in WordPress php side right now (only the impact is more severe because the unclosed tags can wreak havoc on the actual rendered page, whereas with react at least, just the string is mangled.

At this point, I think I should remove the extra block I added, drop the test, and merge this.  If anyone wants to give a go at trying to improve the algorithm (while avoiding any significant performance hits) that can be done in subsequent pulls (being this is experimental that's okay I think).

Thoughts?

Reply:
I'm in favor of removing the block because it starts to look like special-casing and the need cascades beyond where we're at.

The block parser throws invalid output on invalid input. If matching tags are missing then I think we can jumble the output and let that be. Anything to surface the problem sooner will help.

Once we start trying to catch all the errors and _handle_ them then we get into the territory of _hiding_ those problems from another developer by being too enabling and on our side we just have to scratch our heads and wonder why all the extra code that doesn't seem like it should exist is there 🙃